app/services/vine_voucher_validator_service.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+class VineVoucherValidatorService
+  attr_reader :voucher_code, :errors
+
+  def initialize(voucher_code:, enterprise:)
+    @voucher_code = voucher_code
+    @enterprise = enterprise
+    @errors = {}
+  end
+
+  def validate
+    if vine_settings.nil?
+      errors[:vine_settings] = I18n.t("vine_voucher_validator_service.errors.vine_settings")
+      return nil
+    end
+
+    response = call_vine_api
+
+    if !response.success?
+      handle_errors(response)
+      return nil
+    end
+
+    save_voucher(response)
+  rescue Faraday::Error => e
+    Rails.logger.error e.inspect
+    Bugsnag.notify(e)
+
+    # TODO do we need a more specific error ?
+    errors[:vine_api] = I18n.t("vine_voucher_validator_service.errors.vine_api")
+    nil
+  end
+
+  private
+
+  def vine_settings
+    ConnectedApps::Vine.find_by(enterprise: @enterprise)&.data
+  end
+
+  def call_vine_api
+    # Check voucher is valid
+    jwt_service = VineJwtService.new(secret: vine_settings["secret"])
+    vine_api = VineApiService.new(api_key: vine_settings["api_key"], jwt_generator: jwt_service)
+
+    vine_api.voucher_validation(voucher_code)
+  end
+
+  def handle_errors(response)
+    if response.status == 400
+      errors[:invalid_voucher] = I18n.t("vine_voucher_validator_service.errors.invalid_voucher")
+    elsif response.status == 404
+      errors[:not_found_voucher] = I18n.t("vine_voucher_validator_service.errors.not_found_voucher")
+    else
+      errors[:vine_api] = I18n.t("vine_voucher_validator_service.errors.vine_api")
+    end
+  end
+
+  def save_voucher(response)
+    voucher_data = response.body["data"]
+
+    # Check if voucher already exist
+    voucher = Voucher.vine.find_by(code: voucher_code, enterprise: @enterprise)
```

Comment:
We are not validating if the external IDs match up here, I guess we can assume they haven't changed?

The reason I think of this, is because I imagined this could be written without branching by using `find_or_initialize_by`, then setting the amount before saving.

Reply:
> We are not validating if the external IDs match up here, I guess we can assume they haven't changed?

That's a good point, I don't think the external IDs can changes, but I believe it's possible to for the "short code" to be reused. I'll double check that.

Reply:
Paul said :

> short codes are unique but the db field is nulled out once its redeemable value is exhausted
> At that point, the short code might be recycled, but it won't be recycled immediately

I'll need to fix the code to cover the potential recycling of short code

Reply:
I am thinking of doing something like : 
```ruby
  validates :code, uniqueness: { scope: :enterprise_id }, if: ->(voucher) { !voucher.vine? }
  validates :code, uniqueness: { scope: :external_voucher_id }, if: ->(voucher) { voucher.vine? }
```
This would allow VINE to recycle short code but we would need to drop the unique index on code and enterprise_id in the db. 
I think as long we use the appropriate scope when dealing with voucher we should be good ? Am I missing something ? @openfoodfoundation/developers